src/charm.py

```diff
@@ -55,11 +55,17 @@ def __init__(self, *args):
         )
 
         # When a new consumer of the metrics-endpoint relation is related,
-        # pass it all the current scrape jobs
+        # pass it all the current scrape jobs. We register to both the relation
+        # being created and joined, because relation_created is not reliably sent
+        # over unit recreation.
         self.framework.observe(
             self.on[self._prometheus_relation_name].relation_created,
             self._set_jobs_to_new_downstream,
         )
+        self.framework.observe(
```

Comment:
Setting jobs in both relation created and relation joined when one follows the other is a bit surprising. It may be useful to have a comment here explaining why both are needed and in particular why the same event handler is bound to both if both are needed.

Reply:
I think the point here is that the `joined` one is vital, while `created` is a nice bonus - which is opposite to the previous state.

Given that the event handler is (or at least should be) idempotent anyway, running it for both does not really make any difference if everything goes according to plan but also safeguards against any hickups in case an unexpected sequence of events would be emitted.

Reply:
Ack. Will put in a comment to this effect in some future PR.

Reply:
> I think the point here is that the joined one is vital, while created is a nice bonus - which is opposite to the previous state.

The point is that they are both vital for different use-cases ;-)

Reply:
No dispute here, we are alligned. All I was saying was I would be genuinely interested to understand those different use cases. Encoding the experience that led you to this change in a commit message or into comments in code will help future maintainers of this project.

Reply:
For future readers stumbling upon this, i'll quote a couple of lines from the the relation docs:

> ### `<name>_relation_created`
> Triggered when a new relation is created. This can occur before applications have started.
>
> ### `<name>_relation_joined`
> Triggered when a new unit joins a relation. This event only fires when the remote unit is first observed by the unit.
>
> Callback methods bound to this event may set any settings that can be determined by only using the joining unit’s name and the remote private-address setting.

https://juju.is/docs/sdk/relations#heading--event-types